File: src/ipc/normalTable.ts
import { call } from "./client";

export interface NormalTableSchema {
  block: NormalTableBlock[];
}

export type NormalTableBlock =
  | { type: "kv"; items?: NormalKvItem[] }
  | { type: "note"; text?: string }
  | {
      type: "grid";
      title?: string;
      title_style?: string;
      columns?: string[];
      rows?: NormalGridRow[];
    };

export interface NormalKvItem {
  label?: string;
  value?: string;
}

export interface NormalGridRow {
  label?: string;
  cells?: string[];
}

export const getNormalTableSchema = () =>
  call<NormalTableSchema>("get_normal_table_schema");

File: src-tauri/src/config/normal_table.rs
use std::path::Path;

use serde::{Deserialize, Serialize};

use crate::error::AppResult;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct NormalTableSchema {
    #[serde(default)]
    pub block: Vec<NormalTableBlock>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum NormalTableBlock {
    Kv {
        #[serde(default)]
        items: Vec<NormalKvItem>,
    },
    Note {
        #[serde(default)]
        text: String,
    },
    Grid {
        #[serde(default)]
        title: String,
        #[serde(default)]
        title_style: String,
        #[serde(default)]
        columns: Vec<String>,
        #[serde(default)]
        rows: Vec<NormalGridRow>,
    },
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct NormalKvItem {
    #[serde(default)]
    pub label: String,
    #[serde(default)]
    pub value: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct NormalGridRow {
    #[serde(default)]
    pub label: String,
    #[serde(default)]
    pub cells: Vec<String>,
}

impl NormalTableSchema {
    pub fn load(path: &Path) -> AppResult<Self> {
        let text = std::fs::read_to_string(path)?;
        Ok(toml::from_str(&text)?)
    }
}
